warc2zim/main.py

```diff
@@ -0,0 +1,309 @@
+The WARCs are converted in a 'lossless' way, no data from WARC records is lost.
+Each WARC record results in two ZIM articles:
+- The WARC payload is stored under /A/<url>
+- The WARC headers + HTTP headers are stored under the /H/<url>
+
+Given a WARC response record for 'https://example.com/', two ZIM articles are created /A/example.com/ and /H/example.com/ are created.
+
+Only WARC response and resource records are stored.
+
+If the WARC contains multiple entries for the same URL, only the first entry is added, and later entries are ignored. A warning is printed as well.
+
+"""
+
+from argparse import ArgumentParser, RawTextHelpFormatter
+
+from warcio import ArchiveIterator
+from libzim.writer import Article, Blob, Creator
+import os
+import logging
+import requests
+
+# Shared logger
+logger = logging.getLogger('warc2zim')
+
+DEFAULT_REPLAY_SOURCE_URL = 'https://cdn.jsdelivr.net/npm/replaywebpage@1.1.0-alpha.1/'
+
+
+
+# ============================================================================
+class BaseArticle(Article):
+    """ BaseArticle for all ZIM Articles in warc2zim with default settings
+    """
+    def is_redirect(self):
+        return False
+
+    def get_title(self):
+        return ''
+
+    def get_filename(self):
+        return ''
+
+    def should_compress(self):
+        return True
+
+    def should_index(self):
+        return False
+
+
+# ============================================================================
+class BaseWARCArticle(BaseArticle):
+    """ BaseWARCArticle that produces ZIM articles from WARC records
+    """
+    def __init__(self, record):
+        super(BaseWARCArticle, self).__init__()
+        self.record = record
+
+
+# ============================================================================
+class WARCHeadersArticle(BaseWARCArticle):
+    """ WARCHeadersArticle used to store the WARC + HTTP headers as text
+    Usually stored under H namespace
+    """
+    def __init__(self, record):
+        super(WARCHeadersArticle, self).__init__(record)
+        self.url = record.rec_headers.get('WARC-Target-URI')
+
+    def get_url(self):
+        return 'H/' + self.url.split('//', 2)[1]
+
+    def get_mime_type(self):
+        return 'application/warc-headers'
+
+    def get_data(self):
+        # add WARC headers
+        buff = self.record.rec_headers.to_bytes(encoding='utf-8')
+        # add HTTP headers, if present
+        if self.record.http_headers:
+            buff += self.record.http_headers.to_bytes(encoding='utf-8')
+
+        return Blob(buff)
+
+
+# ============================================================================
+class WARCPayloadArticle(BaseWARCArticle):
+    """ WARCPayloadArticle used to store the WARC payload
+    Usually stored under A namespace
+    """
+    def __init__(self, record):
+        super(WARCPayloadArticle, self).__init__(record)
+        self.payload = record.content_stream().read()
```

Comment:
No specific reason, think I just wanted to be sure it was read at a known time.. But since `get_data()` is always called once, should be fine to move there

Reply:
Moved to load from `get_data()`